src/hypothesis/extra/pandas/impl.py

```diff
@@ -70,6 +104,136 @@ def dtype_for_elements_strategy(s):
     )
 
 
+class IndexStrategy(st.SearchStrategy):
+    def __init__(
+        self, elements, dtype, min_size, max_size, unique, order,
+        allow_nan
+    ):
+        super(IndexStrategy, self).__init__()
+        self.elements = elements
+        self.dtype = dtype
+        self.min_size = min_size
+        self.max_size = max_size
+        self.unique = unique
+        self.order = order
+        self.allow_nan = allow_nan
+
+    def do_draw(self, data):
+        result = []
+        seen = set()
+
+        iterator = cu.many(
+            data, min_size=self.min_size, max_size=self.max_size,
+            average_size=(self.min_size + self.max_size) / 2
+        )
+
+        if self.dtype is None:
+            def convert(x): return x
+        else:
+            convert = self.dtype.type
+
+        while iterator.more():
+            elt = convert(data.draw(self.elements(len(result))))
+
+            if not self.allow_nan:
+                try:
+                    if np.isnan(float(elt)):
+                        iterator.reject()
+                        continue
+                except TypeError:
+                    pass
+
+            if self.order != 0 and result:
+                if (
+                    (elt > result[-1] and self.order < 0) or
+                    (elt < result[-1] and self.order > 0)
+                ):
+                    iterator.reject()
+                    continue
+
+            if self.unique:
+                if elt in seen:
+                    iterator.reject()
+                    continue
+                seen.add(elt)
+            result.append(elt)
+
+        if not result and self.dtype is None:
+            dtype = data.draw(dtype_for_elements_strategy(self.elements(0)))
+            return pandas.Index([], dtype=dtype)
+        return pandas.Index(result, dtype=self.dtype)
+
+
+@st.cacheable
+@st.defines_strategy
+def indexes(
+    elements=st.just, dtype=None, min_size=0, max_size=None, unique=True,
+    order=0, allow_nan=False
+):
+    """Provides a strategy for generating values of type pandas.Index.
+
+    * elements is a function which takes an integer position in the index and
+      returns a strategy for generating values. The default will simply
+      generate the current position, converted to the passed dtype if one is
+      provided.
+    * dtype is the dtype of the resulting index. If None, it will be inferred
+      from the elements strategy (and so will default to int64 if neither are
+      passed).
+    * min_size is the minimum number of elements in the index.
+    * max_size is the maximum number of elements in the index. If None then it
+      will default to a fairly small size (currently min_size + 10, but that
```

Comment:
I would omit the explanation, since the distribution is basically an implementation detail.  I'd also make the real maximum size considerably larger (eg min + 100), but keep the average at min + 5 by default.